### PR TITLE
FEATURE: Added Capacity per User Feature

### LINKED ISSUE #3: FEATURE: As an admin, I can specify a capacity "per farmer"
# User Story

* As an admin
* I can specify a commons capacity on a per farmer basis
* so that the game will scale with the number of users automatically

### Discussion

Prof. de Vries would like to be able to specify commons capacity on a "per user" basis.

The idea is to add a field `capacityPerUser` that will be multiplied by the number of users (farmers) in the commons.    

Then, when doing capacity calculations for the health update formulas, the capacity used will be `max(carryingCapacity, numUsers * capacityPerUser)`.

### Acceptance Criteria

- [x] There is a way for the admin to set capacityPerUser in addition to carryingCapacity; and both fields appear when you create or edit a commons.
- [x] On the `/admin/listcommons` page, you can see both the carrying capacity, the perUserCapacity, and the effectiveCapacity, which is `max(carryingCapacity, numUsers * capacityPerUser)`
- [x] The effectiveCapacity is used in update health formulas instead of the carryingCapacity.

### Implementation Todos

# Fix backend crud operations

- [x] The Commons entity has a new field for capacityPerUser
- [x] The POST operation for creating a new commons can set this field
- [x] The PUT operation for editing a commons can update this field
- [x] The commons entity has a getter for effectiveCapacity which return `max(carryingCapacity, numUsers * capacityPerUser)`
- [x] The commons entity has annotation to return the value of this serialized in the JSON.

# Fix health operations
- [x] The health formulas are updated to use effectiveCapacity

# Fix frontend
- [x] Frontend create and edit pages for commons provide access to capacityPerUser
- [x] List Commons page for admins shows capacityPerUser and effectiveCapacity in addition to all current fields.

### PR DESCRIPTION
**Pull request by MPriston**
## Overview

In this PR we add the ability to describe a capacity per-user when creating a new common or editing it. This change introduces a new field on the create and edit commons form that gives the admin the ability to specify a per-user capacity for each commons.

It utilizes the following formula to calculate the effective capacity of the common when doing health calculations
`max(capacityPerUser * numUsers, carryingCapacity)`

### New Create and Update Form:
image

### Effective Capacity with One User in the common
image
### Effective Capacity with Two Users in the common
image

### With 12 cows in the common health updating job does not diminish cow health
image

### With 12 cows in the common, and one user is removed, the health updating job diminishes cow health
image

## Validation (Optional)

1. Download this branch.
2. Run the project.
3. Go to this page.
4. Test this function.

## Feedback Requests
The way I implemented this feature was by creating a new private and Non-serializable variable in the Commons entity, and every time a user joins or is deleted from a common the variable is updated. This was due to the fact that you can't reference repositories inside the entity in order to get the user count. If you have any suggestions on how to improve this solution that will be appreciated.

## Tests

- [x] Backend Unit tests (`mvn test`) pass
- [x] Backend Test coverage (`mvn test jacoco:report`) 100%
- [x] Backend Mutation tests (`mvn test pitest:mutationCoverage`) 100% 
- [x] Frontend Unit tests (`npm test`) pass
- [x] Frontend Test coverage (`npm run coverage`) 100%
- [x] Frontend Mutation tests (`npx stryker run`) 100% 
- [x] Frontend Linting (`npx eslint --fix src`) 

## Linked Issues

Closes #3

### CODE REVIEW

**Comment by MPriston** [comment]
Currently working on the Frontend of the changes

**Comment by MPriston** [comment]
Currently working on some bugs

**Comment by MPriston** [comment]
Working on test coverage for new variables

**Comment by MPriston** [comment]
Test Coverage is good, need to update, jobs notifications to display correct information

File: src/main/java/edu/ucsb/cs156/happiercows/strategies/CowHealthUpdateStrategies.java
Comment (publicqi) [inline src/main/java/edu/ucsb/cs156/happiercows/strategies/CowHealthUpdateStrategies.java:1]: Not your code, but why do we add health if the capacity is reached?

File: src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java
Comment (publicqi) [inline src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java:1]: Is this (`numUsers`) passed in by frontend? If so, we should not trust that.

**Review by publicqi — APPROVED** [review]

**Review by dscpsyl — DISMISSED** [review]
LGTM! Had a team code review in person over this PR.

**Review by publicqi — DISMISSED** [review]

File: frontend/src/fixtures/commonsFixtures.js [hunk 1]
```diff
@@ -41,7 +45,9 @@ const commonsFixtures = {
             "milkPrice": 10,
             "degradationRate": .5,
             "showLeaderboard": true,
+            "capacityPerUser": 13,
             "carryingCapacity": 42,
+            "effectiveCapacity": 42,
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js]: If the commons has 50 players, and the capacity per user is 13, then 13*50=650.   Shouldn't the effective capacity be 650 in that scenario?

File: frontend/src/fixtures/commonsFixtures.js [hunk 2]
```diff
@@ -26,7 +28,9 @@ const commonsFixtures = {
             "milkPrice": 10,
             "degradationRate": .5,
             "showLeaderboard": true,
+            "capacityPerUser": 11,
             "carryingCapacity": 123,
+            "effectiveCapacity": 123,
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js]: If the totalPlayers are 50 and the capacityPerUser is 11, shouldn't the effectiveCapacity be 550?

File: frontend/src/fixtures/commonsFixtures.js [hunk 3]
```diff
@@ -11,7 +11,9 @@ const commonsFixtures = {
             "milkPrice": 10,
             "degradationRate": .5,
             "showLeaderboard": true,
+            "capacityPerUser": 10,
             "carryingCapacity": 100,
+            "effectiveCapacity": 100,
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js]: If the totalPlayers are 50, and the capacityPerUser is 10, shouldn't the effectiveCapacity be 500?

File: frontend/src/fixtures/commonsFixtures.js [hunk 4]
```diff
@@ -74,7 +82,9 @@ const commonsFixtures = {
             "cowPrice": 15,
             "degradationRate": .5,
             "showLeaderboard": true,
+            "capacityPerUser": 50,
             "carryingCapacity": 100,
+            "effectiveCapacity": 100,
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js:87]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsFixtures.js [hunk 5]
```diff
@@ -86,7 +96,9 @@ const commonsFixtures = {
...
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsFixtures.js [hunk 6]
```diff
@@ -98,7 +110,9 @@ const commonsFixtures = {
...
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsFixtures.js [hunk 7]
```diff
@@ -110,7 +124,9 @@ const commonsFixtures = {
...
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js:129]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsFixtures.js [hunk 8]
```diff
@@ -122,7 +138,9 @@ const commonsFixtures = {
...
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js:143]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsFixtures.js [hunk 9]
```diff
@@ -134,7 +152,9 @@ const commonsFixtures = {
...
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js:157]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsFixtures.js [hunk 10]
```diff
@@ -146,7 +166,9 @@ const commonsFixtures = {
...
```
Comment (pconrad) [inline frontend/src/fixtures/commonsFixtures.js:171]: Please check this effectiveCapacity.

File: frontend/src/fixtures/commonsPlusFixtures.js [hunk 11]
```diff
@@ -11,7 +11,9 @@ const commonsPlusFixtures = {
                 "endingDate": null,
                 "degradationRate": 0.01,
                 "showLeaderboard": false,
+                "capacityPerUser": 50,
                 "carryingCapacity": 100,
+                "effectiveCapacity": 100,
```
Comment (pconrad) [inline frontend/src/fixtures/commonsPlusFixtures.js:16]: Please check all of the effectiveCapacities to make sure they are internally consistent.

**Review by pconrad — CHANGES_REQUESTED** [review]
It appears that the fixtures may send future developers mixed signals about the rules governing effectiveCapacity.

**Review by pconrad — CHANGES_REQUESTED** [review]
Check out this segment of code in `src/main/java/edu/ucsb/cs156/happiercows/jobs/UpdateCowHealthJob.java`

Should this be updated?

```java
 int carryingCapacity = commons.getCarryingCapacity();
            Iterable<UserCommons> allUserCommons = userCommonsRepository.findByCommonsId(commons.getId());
            Integer totalCows = commonsRepository.getNumCows(commons.getId()).orElseThrow(() -> new RuntimeException("Error calling getNumCows(" + commons.getId() + ")"));
            var isAboveCapacity = totalCows > carryingCapacity;
```

**Review by pconrad — APPROVED** [review]
Nice job!

**Pull request merged**